akka-actor-typed/src/main/scala/akka/actor/typed/internal/delivery/ProducerController.scala

```diff
@@ -0,0 +1,333 @@
+/*
+ * Copyright (C) 2019 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.actor.typed.internal.delivery
+
+import scala.concurrent.duration._
+import scala.reflect.ClassTag
+
+import akka.Done
+import akka.actor.typed.ActorRef
+import akka.actor.typed.Behavior
+import akka.actor.typed.scaladsl.ActorContext
+import akka.actor.typed.scaladsl.Behaviors
+import akka.actor.typed.scaladsl.TimerScheduler
+import akka.actor.typed.scaladsl.LoggerOps
+
+// FIXME Scaladoc describes how it works, internally. Rewrite for end user and keep internals as impl notes.
+
+/**
+ * The `ProducerController` sends `RequestNext` to the actual producer, which is then allowed to send
+ * one message to the `ProducerController`. The `RequestMessage` message is defined via a factory
+ * function so that the producer can decide what type to use. The producer and `ProducerController`
+ * are supposed to be local so that these messages are fast and not lost.
+ *
+ * The `ProducerController` sends the first message to the `ConsumerController` without waiting for
+ * a `Request` from the `ConsumerController`. The main reason for this is that when used with
+ * Cluster Sharding the first message will typically create the `ConsumerController`. It's
+ * also a way to connect the ProducerController and ConsumerController in a dynamic way, for
+ * example when the ProducerController is replaced.
+ *
+ * When the first message is received by the `ConsumerController` it sends back the initial `Request`.
```

Comment:
I read the implicit as having a deliver->confirmed roundtrip as well before it sends the initial  `Request`, am I mistaken?

Reply:
first Request is sent back immediately, there is one additional Request sent back for the first 1L message after the Confirmation, but the purpose of that is to confirm that the first has been completed to cancel the special resending of the first from PC. I will probably also introduce a Ack message for other reasons, and that can be used for that latter case.